4.1 datasource locatorinterface

ADmad opened the pull request:
## The problem

Let's say someone want to use a custom locator for `Table` class instances throughout their app. The  global way to achieve that is customize the table locator using `FactoryLocator::set()` and set a callable for type `Table`.

Doing so would then make `ModelAwareTrait::loadModel()` (used in `Controller`) to use the callable one has set to get table instances. 

Another way to get table instances is using the `LocatorAwareTrait` (`Controller` uses it) and use `LocatorAwareTrait::getTableLocator()->get()`. Now the problem is `LocatorAwareTrait::getTableLocator()` by default will always return a `TableLocator` instance. So one would have to use `LocatorAwareTrait::setTableLocator()` to set a custom table locator which achieves the same as what the callable used for `FactoryLocator`.

This isn't practical. There should be a way to globally set a custom table locator which is used throughout the ORM as single source for table instances.

## Possible solutions

### Allow the default table locator class in `TableRegistry` to be customizable :-1: 
`LocatorAwareTrait::getTableLocator()` use `TableRegistry::getTableLocator()` to get the table locator. `TableRegistry` currently doesn't provide anyway to customize the default table locator class. So we could add public method to be able to do so.

But again this solution isn't practical. One would then have to configure `FactoryLocator` as well as `TableRegistry`, plus the former uses callable while the latter `ORM\LocatorInterface` instances. So this isn't much of a solution. Moreover `TableRegistry` is essentially a relic for BC compatibility which should be removed in future.

### Add a `Datasource\LocatorInterface` and allow `FactoryLocator` to use both callable and the interface instances. :+1: 

This is the solution I have implemented here.

- Added `Datasource\LocatorInterface` (there's also another issue for this)
- Updated `FactoryLocator` to use both callable and `Datasource\LocatorInterface` instances.
- `ORM\LocatorInterface` now extends `Datasource\LocatorInterface`.
- Updated the codebase to always use `FactoryLocator` as the single consistent source for table locator instances.

## Implications
- To allow `ORM\LocatorInterface` to `Datasource\LocatorInterface` the `Table` typehint for `$object` argument of `ORM\LocatorInterface::set()` had to be removed. This is technically a BC break and will throw an error for anyone who has extended `TableLocator` and customized the method. But IMO that's acceptable and anyone using a custom locator would actually appreciate this change, to be able to customize table locator from a single location.

If the changes seem reasonable I'll go ahead with making static analyzers happy and any other related potential clean ups that can be done.

ADmad committed:
Add Datasource\LocatorInface.

ORM\LocatorInterface now extends the above interface.

ADmad committed:
Allow FactoryLocatory to use Datasource\LocatorInterface instances.

ADmad committed:
Use FactoryLocator as the single common source to get table locator.

dereuromark commented on src/ORM/Locator/TableLocator.php:
```diff
@@ -274,7 +274,7 @@ public function exists(string $alias): bool
      * @inheritDoc
      */
-    public function set(string $alias, Table $object): Table
+    public function set(string $alias, $object): Table
```
Could this be `object` ?

ADmad replied:
Yes, `object` typehint can be used that argument here (and in the interfaces).

ADmad committed:
Fix typehint and update errors reported by psalm

ADmad committed:
Allow ModelAwareTrait::modelFactory() to accept Datasource\LocatorInterface instance.

ADmad committed:
Use null coalescing operator

ADmad committed:
Update FactoryLocator tests.

Removed duplicated tests for ModelAwareTrait::loadModel().
These tests are already present in ModelAwareTraitTest class.

CauanCabral commented:
Could that changes also remove `Table` as the return typehint?

That way, others datasources, like `CakePHP\ElasticSearch` can utilize the same interfaces and traits to load its repositories.

markstory commented on src/Datasource/FactoryLocator.php:
```diff
@@ -27,18 +27,18 @@ class FactoryLocator
      * @return void
      */
-    public static function add(string $type, callable $factory): void
+    public static function add(string $type, $factory): void
```
Should we do checks here to ensure that $factory is of an acceptable type?

ADmad replied:
Done

ADmad commented:
> Could that changes also remove Table as the return typehint?

> That way, others datasources, like CakePHP\ElasticSearch can utilize the same interfaces and traits to load its repositories.

The new `Cake\Datasource\LocatorInterface` does not use `Table` typehint. Only `Cake\ORM\Locator\LocatorInterface` and `Cake\ORM\Locator\LocatorAwareTrait` which are under the `ORM` package use `Table` typehint for return value.

The elastic search's `Cake\ElasticSearch\IndexRegistry` would implement `Cake\Datasource\LocatorInterface`. It already all the required methods except `getConfig()` and `setConfig()`. Besides that just the `Index` typehint for the argument of `IndexRegistry::set($alias, Index $object)` would have to be removed.

ADmad committed:
Validate argument types.

ADmad commented:
The implementations of methods like `setConfig()`, `getConfig()`, `exits()` of `LocatorInterface` will generally be identical across all classes. So I am wondering if providing a base/abstract class which implements these methods would be nice to avoid code duplication.

ADmad committed:
Move Datasource\LocatorInterface to Datasource\Loctor\LocatorInterface

ADmad committed:
Add Datasource\Locator\AbstractLocator

Classes implementing Datasource\Locator\LocatorInterface can extend
this class to avoid code duplication.

dereuromark commented on src/ORM/Locator/TableLocator.php:
```diff
@@ -27,7 +28,7 @@
  * Provides a default registry/factory for Table objects.
  */
-class TableLocator implements LocatorInterface
+class TableLocator extends AbstractLocator implements LocatorInterface
```
The interface is already implemented and should be removed here.

ADmad replied:
This is `ORM\LocatorInterace` not `Datsource\Locator\LocatorInterface`.

dereuromark replied:
Hah ok :) My bad.

ADmad commented:
I have added a `Cake\Datasource\Locator\AbstractLocator` with common methods, but I now question the usefulness of the `setConfig()` and `getConfig()` method. 

They are meant to set(and get) per alias default config which are merged with `$options` passed to `get()` calls. Do people actually use this feature in their apps? I personally have always passed all required options to `get()` calls. (Also the naming of args `$options` vs `$config` and properties `$_config` vs `$_options` is quite confusing.)

Elasticsearch's `IndexRegistry` doesn't have these `setConfig()` and `getConfig()` methods.

So I think having these for the new `LocatorInterface` under `Datasource` is not necessary. We can retain they only for `ORM\LocatorInterface` for backwards compatibility.

markstory reviewed (approved):
Looks good to me.

ADmad committed:
Remove getConfig(), setConfig() for Datasource's LocatorInterface.

Very few would need this feature.

ADmad commented:
I removed `getConfig()`, `setConfig()` from `Cake\Datasource\Locator\LocatorInterface` and updated `AbstractLocator`.
